Module shapes for easier tooling

voodoos opened the pull request:
## Presentation

This PR extends the typechecker to compute (and store) a secundary representation of a structure, where modules keep their shape (i.e. functor or structure), but every other kind of item is trivialized. To each node in this structure we optionally attach a `Uid.t` (which the typechecker already produces for every binding).

Reductions on this simplified representation have the nice property of propagating uids in the "expected" way, making it easier for external tools (e.g. merlin, rotor, etc.) to answer questions like "_where is X defined ?_". Which in turns simplifies listing all the occurences of a definition in a codebase, as well as implementing refactoring tools.

We call these _shapes_ and they are dumped in a new kind of cm-files: `.cms(i)` The `.cms` files should be able to replace entirely `.cmt` files for Merlin usages. As these files are lighter than cmt files we believe that this should speed up the process of building a program in dev mode, thus improving the speed of small iterations when coding.

This PR is the result of a collaborative work between @voodoos, @trefis and @lpw25 (who mainly took part in the design process).

## The calculus

Shapes are terms of a small untyped calculus:
```
t = shape := {uid: Uid.t option; desc: desc}

item := name * namespace (value, type, module...)
var  := ident

desc := Leaf
      | Var       of var
      | Abs       of var * t
      | App       of t * t
      | Struct    of (item, t) Map.t
      | Proj      of t * item
      | Comp_unit of string
```

- The `Struct` constructor holds a map from items in a structure to their corresponding shape. To access the shape of a specific item in a structure one must use a projection `Proj`. 
- Functors are modeleld using the `Abs` constructor to abstract over their argument and functor application is denoted by `App`. 
- Shapes of external compilation units (which we sometimes project from, or pass to functors) are lazily represented by `Comp_unit` terms.

Reduction rules for this calculus are as follows:
```
App(Abs(var, body), arg) -β-> body[var/arg]

Proj(Struct item_map, item) -π-> Item.Map.find item_map item

Comp_unit name -l-> Load shape of [name]
```
N.B. the first two rules are applied eagerly by the compiler whilst building shapes, whereas the `l` rule is never used by the compiler, only by external tools.

For example, the following structure:
```ocaml
let x = 0

module M = Stdlib.String

module F (X : sig val x : int end) = struct
  include X
end

module N = F (struct let x = x end)
```

Will have a shape of the form (ignoring the uids):
```
{
  "x" -> Leaf
  "M" -> Proj(Comp_unit "Stdlib", "String")
  "F" -> Abs(X, { "x" -> Proj(X, "x") })
  "N" -> { "x" -> Leaf }
}
```

A zoo of example shapes can be found in the new `testsuite/tests/shape` directory provided by this PR.

### How does this help tooling

Let's look at an example extracted from Merlin's testsuite:
```ocaml
module type S = sig type t end

module Identity(X : S) : S = X

module Simple = struct type t end

module Apply(Id : S -> S)(X : S) = struct include Id(X) end

module M = Apply(Identity)(Simple)

type t = M.t
```

Previously, to jump to the definition on `M.t` merlin (and other tools) would have to first find the definition `M`, then look inside it to find `t`. Here "inside it" implies looking inside the definition of `Apply`, realising that `t` is included from an application of its first parameter, looking inside that, then coming back inside apply to find what argument was given when applying the first parameter (it's the second!), finally poping back up to where `Apply` was applied, and ultimately finding t in `Simple`.

With the work from this PR, one simply has to build the shape corresponding to `M.t`, i.e. `Shape.proj (Env.find_shape M) (Item.type_ "t")`, and to call `Shape.reduce` on it. That gives us a `Leaf` with its uid attached; uid which we can then easily translate back to a source location.

## Changes made by this PR

- A new `Shape` module containing the shapes AST and functions to manipulate it. This module also embeds modules `Types.Uid` and `Typemod.Sig_component_kind`.
- Shapes are first computed in `Typemod`, and later refined in `Includemod`.
- Shapes, including intermediary ones, are stored in the environment. This is necessary necessary for Merlin to be able to jump from nested structures of the current compilation unit.
- We also store in `Env` a table assigning `Location.t`s to `Uid.t`s. This table is necessary to remove the dependency of tooling towards `cmt` files. It is then included in the new `cms` files.
- We are adding two command line flags `-shapes` to instruct the compiler to write `.cms[i]` files, and `-dshapes`, for the same reasons as other `-d` flags.
- A handful of tests have been added to the testsuite.

## Benchmarks and tests

Apart from the additions this PR makes to the compiler testsuite, we also have a branch of Merlin where jump-to-definition is reimplemented as discussed above. This new implementation fixes [a couple of known-bugs](https://github.com/ocaml/merlin/commit/de0cdbb390dc251dab0cbae6f4da094854a2a52a).

We also called both versions of jump-to-definition on various random points of the merlin codebase and confirmed the result was either the same or better than before.

As for the benchmarks, they are included for two reasons:
- reassure people that computing these shapes is not remarkably costly
- confirm our intuition that using .cms instead of .cmt for development cycles results in faster (and smaller) builds.

All benchmarks ran in WSL on a 6 core processor.

- The `make world.opt` benchmark consists in building the compiler 10 times with both `cmt` and `cms` files enabled
- The `dune build @install` benchmark consists in building 5 times [a monorepo](https://github.com/emillon/js-monorepo) with a selection of janestreet packages. For 4.13+shapes we used a modified version of Dune that does not output cmt files.

### 4.13

| Command | Mean [s] | Min [s] | Max [s] |
|:---|---:|---:|---:|
| `make world.opt` | 244.678 ± 0.543 | 244.091 | 245.801 |
| `make world.opt -j12` | 69.882 ± 0.944 | 67.876 | 70.965 |
| `dune build @install -j12` | 88.180 ± 0.504 | 87.343 | 88.698 | 

Size of the `_build` dir of the monorepo:

| Element | Size |
|:---|---|
| `_build/default` | 1.61 go |
| `cmt` files | 253mo |
| `cmti` files | 62mo |

### 4.13+shapes (7cd6de6)

| Command | Mean [s] | Min [s] | Max [s] |
|:---|---:|---:|---:|
| `make world.opt` | 247.268 ± 0.395 | 246.618 | 247.882 |
| `make world.opt -j12` | 70.654 ± 0.771 | 69.302 | 72.197 |
| `dune build @install -j12` | 82.435 ± 0.582 | 81.459 | 82.898 | 1.00 |

Size of the `_build` dir of the monorepo:

| Element | Size |
|:---|---|
| `_build/default` | 1.36 go |
| `cms` files | 65mo |
| `cmsi` files | 2.6mo |

Octachron commented on typing/shape.ml:
```diff
@@ -0,0 +1,340 @@
+
+let for_persistent_unit s = { uid = None; desc = Comp_unit s }
+
+let leaf_for_unpack = { uid = None; desc = Leaf }
```
I would have expected this leaf to be more specialized since it essentially means that the underlying shape is unknown?

trefis replied:
More specialized how? The underlying shape is indeed unknown but because we really don't know it, not because we're forgetting to build it.

The idea of having a `leaf` here is that anyone trying to jump to definition of a projection from an unpack should just be brought to that unpack.

Octachron replied:
I was thinking of signaling to the user that it is not possible to follow a definition through an unpack, but simply pointing to the unpack might be enough.

Octachron commented:
Concerning the benchmarks, if the 4.13+shapes is emitting both `cms` and `cmt` files, it seems that the comparison with the vanilla 4.13 which is only producing `cmt` files is slightly biased. It might better to have both versions do the same amount of "productive work" (vanilla with cmt vs shapes with cmt, or vanilla with cmt vs shapes with cms).

gasche commented:
From a glance: the idea of shapes is nice and simple, and the benchmarks are reassuring that there is not much overhead. I'm not sure about the idea of storing shapes in yet another place, and about the idea of encouraging to stop installing `.cmt` files.

Naive questions:
- Shapes are presented as just one grammar, with an implementation/interface distinction. What is the difference between `.cms` and `.cmsi`?
- What would Merlin use `.cmt` files for? Is it only for jump-to-definition?

gasche commented on file_formats/cms_format.ml:
```diff
@@ -0,0 +1,61 @@
+
+let read_magic_number ic =
+  let len_magic_number = String.length Config.cms_magic_number in
+  really_input_string ic len_magic_number
```
it would be nice to have support for cms magic numbers in `Misc.Magic_number`.

voodoos replied:
done in e201eca

trefis commented:
> Concerning the benchmarks, if the 4.13+shapes is emitting both cms and cmt files, it seems that the comparison with the vanilla 4.13 which is only producing cmt files is slightly biased. It might better to have both versions do the same amount of "productive work" (vanilla with cmt vs shapes with cmt, or vanilla with cmt vs shapes with cms).

Actually IIRC the description of that benchmark is slightly misleading (sorry!): `.cms` were produced only for the stdlib (so merlin can jump to stdlib files), but not for the rest of the distribution; so it's actually closer to what you describe.

Anyway, since we intend to turn `.cms` generation on by default, it seems reasonable to me to compare the time needed for the "default" build of the distribution.

> I'm not sure about the idea of storing shapes in yet another place, and about the idea of encouraging to stop installing `.cmt` files.

Note: we're very much not trying to discourage installing `.cmt` files, they are still needed by odoc for instance (and possibly other tools I don't know about).

What we're trying to change is the necessity of building them during dev builds. `.cmt` are big, and slow to build. It might not matter for small projects, but my memories from janestreet tell me that it will matter for big repositories like theirs.
Perhaps @lpw25 (or @sliquister) could weigh in?

> What would Merlin use `.cmt` files for? Is it only for jump-to-definition?

Yes, thus far `.cmt` have only ever been used for jump-to-definition.

> Shapes are presented as just one grammar, with an implementation/interface distinction. What is the difference between `.cms` and `.cmsi`?

Not sure what you mean by "implementation/interface distinction". We only build shapes for implementations. Uids for _declarations_ (i.e. interfaces) are already propagated as they should by the typechecker.

So a "jump-to-declaration" based on uids could be written today even without this PR (and in fact, we've prototyped one in merlin).
But for "jump-to-definition" shapes are required.

As for what `.cmsi` are used for if there are no shapes for interfaces: they simply store the table mapping uids to locations.

trefis commented:
Btw, one thing we have considered is to save shapes in `.cmt` files as well as in `.cms` file. The rational being that while dune will probably quickly update to build (and install) `.cms` files, other projects will other build systems might be slower to update.
It would be nice for merlin to keep working with those projects.

What's left implicit in that message is that we fully intend to remove the thousands of line in merlin that currently (slightly incorrectly) browse through `.cmt`s. But it would be acceptable to load `.cmt` files to get shapes in the absence of a `.cms`.
(But again: that would just be a fallback mode, I really think not having to always build cmts will make the world a better place)

gasche commented:
> Note: we're very much not trying to discourage installing .cmt files, they are still needed by odoc for instance (and possibly other tools I don't know about).

If that is the intent, then benchmarking `dune build @install` with an experimental version of Dune that does not build `.cmt` files as part of `@install` is a bit confusing :-)

> Not sure what you mean by "implementation/interface distinction".

I meant "without an implementation/interface distinction" instead of "with", apologies. Basically my question is: for a compilation unit `foo.ml` and `foo.mli`, what goes in `foo.cms` and `foo.cmsi`?

My understanding is that `.cms` store the shape computed from the implementation `foo.ml`. Do I correctly understand from your reply that `foo.cmsi` is used to store the mapping from uids in `foo.cms` to locations? (Why not store the mapping in `foo.cms` and have a single kind of file?) Is there the idea somewhere of computing the restriction of the shape of `foo.ml` by only the interface exposed by `foo.mli`?

trefis commented:
Right!

So let's make things clearer:
- `foo.cms` contains: the shape for `Foo`, the table mapping uids from `foo.ml` to locations in that file
- `foo.cmsi` contains: the table mapping uids from `foo.mli` to locations in that file

>  Is there the idea somewhere of computing the restriction of the shape of `foo.ml` by only the interface exposed by `foo.mli`?

Yes! That is done by `Includemod` at the same time as the computation of the module coercion (that's what was meant by "later refined by Includemod" in the original message).

---

> benchmarking dune build @install with an experimental version of Dune that does not build .cmt files as part of @install is a bit confusing :-)

Fair! That patched dune is more a "hack" than an experimental version, it was the quickest way we found to measure what we wanted :)

dbuenzli commented:
> (But again: that would just be a fallback mode, I really think not having to always build cmts will make the world a better place)

What about avoiding adding more compilation object types and provide a flag where you don't include in `.cmt`s these things that make the world such a horrible place ?

gasche commented:
> * `foo.cms` contains: the shape for `Foo`, the table mapping uids from `foo.ml` to locations in that file
> 
> * `foo.cmsi` contains: the table mapping uids from `foo.mli` to locations in that file

Would it make sense to store the mapping from `Uid.t` to locations in the `.cmi` and `.cm[ox]` files directly (this is, morally, type-checking-time debugging information), and only have `foo.cms` that stores the shape?

Another naive question, sorry:

> Previously, to jump to the definition on `M.t` merlin (and other tools) would have to first find the definition `M`, then look inside it to find `t`. Here "inside it" implies looking inside the definition of `Apply`, realising that `t` is included from an application of its first parameter, looking inside that, then coming back inside apply to find what argument was given when applying the first parameter (it's the second!), finally poping back up to where `Apply` was applied, and ultimately finding t in `Simple`.

I would think of a more direct route, which is of calling `Env.find_type "M.t" current_env`, and getting the uid from the `type_uid` field of the returned `type_declaration`. Why does this approach not work?

voodoos commented:
> I would think of a more direct route, which is of calling `Env.find_type "M.t" current_env`, and getting the uid from the `type_uid` field of the returned `type_declaration`. Why does this approach not work?

The uid you get from the `type_declaration` record is the one of the declaration of `M.t` not its definition. In fact we follow that exact route to perform jump-to-declaration in our new Merlin implementation, but shapes are needed to find the definition's uids.

trefis commented:
> What about avoiding adding more compilation object types and provide a flag where you don't include in `.cmt`s these things that make the world such a horrible place ?

I'm not keen on the idea of selectively putting things in cmts.
And from a build system perspective it seems rather annoying: there's this one file which will be produced differently depending on who asks for it? :grimacing: 

> Would it make sense to store the mapping from Uid.t to locations in the `.cmi` and `.cm[ox]` files directly (this is, morally, type-checking-time debugging information), and only have foo.cms that stores the shape?

I'd be happy to store the table in `.cmi`s to make `.cmsi` disappear if people are OK with that idea.
But if `.cms` files are still present, why not save the implementation table there, instead of in `.cm[ox]`? It doesn't matter much, but on principle I'm a bit reluctant to make merlin depend on anything "from the backend".

dbuenzli commented:
> I'm not keen on the idea of selectively putting things in cmts.
> And from a build system perspective it seems rather annoying: there's this one file which will be produced differently depending on who asks for it? 😬

Well that's what for example `-g` does…

mshinwell commented:
> I'd be happy to store the table in `.cmi`s to make `.cmsi` disappear if people are OK with that idea.

Fewer different varieties of artifact presumably helps to keep the computation graph inside the build system smaller (which is a good thing)?

lpw25 commented:
The compiler loads a lot of cmi files. Adding more to them might affect performance. I'd rather have the separate cmsi file, and consider it a replacement for cmti files in the usual dev build.

gasche commented:
> > I would think of a more direct route, which is of calling `Env.find_type "M.t" current_env`, and getting the uid from the `type_uid` field of the returned `type_declaration`. Why does this approach not work?
> 
> The uid you get from the `type_declaration` record is the one of the declaration of `M.t` not its definition. In fact we follow that exact route to perform jump-to-declaration in our new Merlin implementation, but shapes are needed to find the definition's uids.

Thanks for the clarification. So, if I understand correctly, when you write

> Reductions on this simplified representation have the nice property of propagating uids in the "expected" way, making it easier for external tools (e.g. merlin, rotor, etc.) to answer questions like "where is X defined ?"

There are in fact two form of Uid.t propagations:
- uids currently stored in the typing environment point to the *declaration point* of the item
- uids stored in the (new) shapes point to the *definition point* of the item

Some other naive questions follow:
1. Are there more than one uid we could care about for each item? In particular, if "shapes" are to be a flexible, easy-to-use, informative structure, could/should they consider storing both the definition and the declaration uids? more?
2. Conversely, if we stored definition-uids in the environment in addition to the declaration-uids (and I guess in module coercions as well, etc.), would that be a valid alternative to having shapes to complement interfaces? Or is there more information we can extract from shapes, or is it easier/cheaper to propagate definition-uids in shapes than in signatures, or something else?

lpw25 commented:
I think a key point to observe is that the declaration information flows through typing and comes from `cmi` files, whilst the definition information depends on the actual implementation of the module, which is not available when the `cmi` file is created (assuming there is an `mli` file). This is why other compilation units must be treated as free variables in the shapes: when you create the shape of a unit it cannot yet see the shapes of the units it depends on.

samoht commented:
Am I correct to see these `cms` files are "lighter" `cmt` files? e.g. we introduced `cmt` a few years ago as a replacement for `.annot` files to be used by editors.  Now we make that format evolve to also carry extra information. If that's the case, could we pack both `cms` as a new entry in `cmt` and have 2 simple compilation modes for build systems:

- a "dev" mode, where we pass `--compact-cmt` to `ocamlc`, which produces `cmt` files with only what you propose to put here in `cms` file.
- a "profile" mode, where we pass `--fat-cmt` to `ocamlopt`, which produces all of what is in `cmt` today + the extra new information that you want to put in `cms`.  

We could pick the default that we want (probably `--fat-cmt` to not break existing tooling), but that might simplify the integration with existing tooling.

Octachron commented:
I had a diagram drawing session with @trefis to see how shapes fit among the typed tree (aka the cmt files) and the module type (aka the cmi files). For interfaces, we have:
```
Typed tree
   │      ╲
   │       ╲ 
   ∇        ◿ 
  Types────⊳Shape
```
In words, the shape is a simplification of the type which is itself a simplification of the typed tree. Thus, it makes sense to store only the location of the UIDs in the `cmsi` file since the type declaration are easily available in the cmi file. Which make me wonder, if the cmsi file is storing meta information away from the main cmi file, would we also want to exfiltrate locations and documentation information from the cmi file to the cmsi file?

Contrarily for implementation, we have approximately:
```
Typed tree ───────⊳ Typed tree without
   │     │  ╲         signature constraints
   │     │   ╲             │
   │     │    ╲            │
   ∇     │     ╲           ∇
 Types   │       ───⊳ Types without
          ╲           signature constraints
           ╲               │
            ╲              │
             ╲             ∇
               ────────⊳ Shape
```
where the important part is that we don't have an arrow anymore between the module type and the shape. Indeed, signature constraints block the flow of information for type declaration in module types. In other words, shapes and `.cms` file are still a simplification of the typed tree and `.cmt` file, but `.cms` files are no longer derivable from a potential `.cmi` file for implementation.

Looking at this diagram, I am also wondering if would not be clearer to think of `cms` files as simplified `cmi` file but for module implementation.

lpw25 commented:
> I am also wondering if would not be clearer to think of cms files as simplified cmi file but for module implementation

It depends on exactly what you mean, but I think this is really the wrong way to think about module shapes. Shapes are entirely about the actual runtime value of a module, whilst cmi files are about the type of a module. I think a useful example to have in mind is something like:

```ocaml
module M : sig
  include Foo.S
end = struct
  include Bar
end
```
The interface of `M` comes from `Foo` whilst the shape of `M` will come from `Bar`. There is no reason that the way a user constructs the type of a module should be related to the way the user constructs the module itself.

I think part of the confusion here is that the `cmsi` file does not contain any shapes. In fact, I think operations which use the `cms` files and operations which use the `cmsi` files are mutually exclusive. The `cms` file is used for `jump-to-definition` that tries to take you to the definition of the runtime value corresponding to a use of an identifier. The `cmsi` file is used for `jump-to-declaration` that takes you to the origin of the type corresponding to a of an identifier.

Octachron commented on typing/typemod.ml:
```diff
@@ -1786,13 +1774,17 @@ and transl_recmodule_modtypes env sdecls =
+             md_uid }
+         in
+         let id_shape =
+           Option.map (fun id -> id, Shape.var md_uid id) id
```
Why is the shape handled differently than the full module here? As stated in the test suite, the fuel-based reducer defined by `Shape.Make_reduce` cannot handle cases where the type definition is imported from the signature
and will loop on
```ocaml
module rec A: sig type t = C end = A
```
or points to the partial type definition in
```ocaml
module rec A: sig
  type t = int * float
end = struct
  type t = A.t
end
```
Are there obstacles to compute the shape by refining a shape approximated from the signature?

trefis replied:
One issue is that, currently, the approximated signature doesn't contain meaningful uids. So any computation you're going to make on a shape constructed from it is going to be incorrect / useless.

You could of course preallocate uids, and then reuse them for the various iterations of the signature translation.

But I guess the important question is: what should happen of your second example? When asking for the definition of `A.t` should we be brought to line 2 or to line 4?
Our current stance is that it should bring you to line 4, because that's where the definition is. Using a shape constructed from the signature is just not going to get you there. Therefore our placeholder variable shape, that is then looked up during reduction (which happens after the whole recursive block has been typed, and proper shapes computed).

trefis replied:
Btw, following your comment I looked at this part of the code again, and it looks like it was slightly incorrect: the environment was never updated to contain the proper shapes after we compute them.
@voodoos is going to send a patch fixing that soon (i.e. in the next hour or so).

voodoos replied:
Patch is 80d4c6d

trefis replied:
Btw: ping @gasche .
tl;dr: the answer to [your question](https://github.com/ocaml/ocaml/pull/10718#issuecomment-954792476) is: no, we use fuel to stop looping forever on your example.

Octachron replied:
To summarize my discussion with @trefis , currently shapes do not support "definition out-of-declaration" where the implementation definition is imported from the signature. This seems like an acceptable compromise since there is no concrete definitions in those instances.

Octachron commented on driver/main_args.mli:
```diff
@@ -107,6 +108,7 @@ module type Compiler_options = sig
   val _with_runtime : unit -> unit
   val _without_runtime : unit -> unit
   val _safe_string : unit -> unit
+  val _shapes : unit -> unit
```
The `shape` flag will need to be documented in both the manual (see `manual/src/cmds/unified-options.etex`) and the manpage `man/ocamlc.1` and `man/ocamlopt.1` (the filenames changed recently).

Moreover, the name `-shapes` seems potentially uninformative. I think that the name should reflect the fact that this a meta-information/annotation file. Maybe more `-annot-shapes` ?

voodoos replied:
Adressed in 59edafd

voodoos replied:
Patch is 80d4c6d

Octachron commented on typing/shape.ml:
```diff
@@ -0,0 +1,340 @@
+        in
+        Format.fprintf fmt "{@[<v>%a@,%a@]}" print_uid_opt uid print_map map
+  in
+  Format.fprintf fmt"@[%a@]@." aux
```
My opinion is that type printers should not use `@.`: otherwise it is impossible to use the printer in a non-terminal position because `@.` resets all boxes and tags.

voodoos replied:
Done in 7f51c5e

Octachron commented on typing/env.ml:
```diff
@@ -661,7 +679,7 @@ let check_shadowing env = function
   | `Class (Some _) -> Some "class"
   | `Class_type (Some _) -> Some "class type"
-  | `Constructor _ | `Label _
+  | `Constructor _ | `Label _ | `Shape _
```
This `` `Shape `` is the only ocurrence of `` `Shape _ `` in the compiler and looks like a forgotten remnant.

gasche replied:
Adding an explicit closed-variant annotation (here and on the producer as well) would help catch similar mistakes in the future.

voodoos replied:
Done in 59bc9c5

Octachron commented on typing/shape.ml:
```diff
@@ -0,0 +1,340 @@
+
+let dummy_mod = { uid = None; desc = Struct Item.Map.empty }
+
+let of_path ~find_shape ?(ns = Sig_component_kind.Module) =
```
My impression is that `ns = Sig_component_kind.Module` is more a most common argument than a default argument. A labeled  argument might be a better fit?

voodoos replied:
Adresed in 4921eb3

Octachron commented on typing/includemod.ml:
```diff
@@ -735,18 +784,24 @@ and modtype_infos ~in_eq ~loc env ~mark subst id info1 info2 =
-    else Some (modtypes ~in_eq:true ~loc env ~mark Subst.identity mty2 mty1)
+    else
+      Some (
+        modtypes ~in_eq:true ~loc env ~mark Subst.identity
```
Beware when rebasing: this should be `~mark:(negate mark)`.

voodoos replied:
Done in befd144

Octachron commented on typing/includemod.mli:
```diff
@@ -147,8 +147,8 @@ val is_runtime_component: Types.signature_item -> bool
-  loc:Location.t -> Env.t -> mark:mark ->
-  module_type -> module_type -> module_coercion
+  ?shape:Shape.t -> loc:Location.t -> Env.t -> mark:mark ->
+  module_type -> module_type -> module_coercion * Shape.t option
```
The invariant that the shape result is None if and only there was no `shape` provided should be documented (since it is used extensively).
Similarly, I think it would be a good idea to have a `modtypes_with_shape` function to keep a way to avoid `Option.get` when we clearly know that we had a shape at hand.

voodoos replied:
Adressed in 8f54ae1

Octachron commented on typing/typemod.ml:
```diff
@@ -1983,13 +1976,13 @@ let check_recmodule_inclusion env bindings =
           try
             Includemod.modtypes ~loc:modl.mod_loc ~mark:Mark_both env
-              mty_actual' mty_decl'
+              mty_actual' mty_decl' ~shape
```
An `Includemod.modtypes_with_shape` would avoid the subsequent call to `Option.get`.

voodoos replied:
Adressed in 8f54ae1

Octachron commented on typing/typemod.ml:
```diff
@@ -2067,23 +2060,23 @@ let package_subtype env p1 fl1 p2 fl2 =
     mod_env = env;
     mod_attributes = [];
-    mod_loc = arg.mod_loc }
+    mod_loc = arg.mod_loc }, shape
```
Since this function is always called with an explicit shape argument or without the optional argument, I would suggest to keep the current code and then define a `wrap_constraint_with_shape` and `wrap_constraint` functions that takes care properly of the two paths.

voodoos replied:
Adressed in 8f54ae1

Octachron commented on typing/env.mli:
```diff
@@ -353,7 +358,14 @@ val enter_cltype:
+(* Same as [enter_signature] but also extends the shape map ([parent_shape])
+   with all the the items from the signature, their shape being a projection
+   from the given shape. *)
+val enter_signature_shape: scope:int -> parent_shape:Shape.Map.t ->
```
Maybe `enter_signature_and_shape` would be more explicit?

voodoos replied:
Done in 85bde1f

Octachron commented on typing/typemod.ml:
```diff
@@ -1812,14 +1804,14 @@ and transl_recmodule_modtypes env sdecls =
          md_attributes=pmd.pmd_attributes}
       in
-      tmd, md.md_uid
+      tmd, md.md_uid, Option.map snd id_shape
```
Rather than keeping only the shape from `id_shape` and asserting later that you always have a shape when `tmd` contains an `id`, why not propagate the full `id_shape` to `type_structure`?

voodoos replied:
We changed the usage in 80d4c6d

gasche commented:
Another idle question: how are the shapes of mutually-recursive modules computed? Is the language of shapes still normalizing in presence of recursive modules?

```ocaml
module rec A
  : sig val x : int -> int end
  = struct include B end
 and B
  : sig val x : int -> int end
  = struct include A end
```

Octachron commented on toplevel/native/topeval.ml:
```diff
@@ -214,8 +214,11 @@ let execute_phrase print_outcome ppf phr =
       in
-      let (str, sg, names, newenv) = Typemod.type_toplevel_phrase oldenv sstr in
+      let (str, sg, names, shape, newenv) =
+        Typemod.type_toplevel_phrase oldenv sstr
```
Self-note: this commit does not compile independently.

gasche commented:
One thing that took me a bit of time to understand is that shapes are abstractions of modules, not of module types. (Two modules with the same module type may have different shapes.)

(Here "abstraction" can be taken in the sense of "erasure": there is an erasure function from program terms to shapes, shapes are not determined by a classification relation like types.)

Compilation units are abstracted into open shapes / shapes with variables. When a project is fully built, we can resolve all shape variables, and therefore normalize shapes to "closed shapes" without shape variables. Consuming tools perform this normalization on-demand.

The fact that shapes are computed during type-checking is somewhat of a distraction. They could also be computed from the final typedtree. (They cannot be computed on untyped terms, as they need to precisely resolve module identifiers, and they cannot be computed on the Lambda form which lost the module structure.)

Octachron commented on typing/typemod.ml:
```diff
@@ -2427,58 +2439,93 @@ and type_structure ?(toplevel = false) funct_body anchor env sstr =
+            | _ -> assert false
+          )
+          shape_map
+          items
```
There is a small discrepancy with the class side here: the uids of private row type variable, which does not correspond to an user definition are registered, whereas for classes (and class types), the code is careful to only register class and class type definitions.

trefis replied:
Fair point, I had missed the fact that regular type declarations generated more items than present in the source!

Actually, I think our handling of classes is not quite right: we should register the various items in our shapes, since they can appear in the source.
Which raises the question: can the `...#row` types generated here be adressed/referred in the source? That will decide whether we want to store them in the shape map or not; but for now I haven't been able to answer (my guess is: no).

Octachron replied:
I don't think it is possible to obtain a syntactic item that points to a private `...#row` type, at the very least this is not intended to be possible.

voodoos replied:
Ok, I have done that in fd1878b

gasche commented:
Note: my impression remains that `.cmsi` are not really a requirement; they do not store shapes, and they only store a mapping for objects that already exist in the `.cmi` to locations -- `.cmi` also already store locations. I think that the corresponding information could be put in the `.cmi`, reducing the number of different artifact types and thus the complexity for everyone.

(One downside of extending `.cmi` in general is that this can hurt separate compilation: the more information in the `.cmi`, the more recompilation is necessary. But is this an issue here? My understanding is that `.cmi` already stores locations, just not as a direct mapping, so something that would change the Uid->location mapping (for the Uids in the cmi) would already change the `.cmi`, at least in `-keep-locs` mode.)

Octachron commented on typing/shape.mli:
```diff
@@ -0,0 +1,140 @@
+  val add : t -> Item.t -> shape -> t
+
+  val add_value : t -> Ident.t -> Uid.t -> t
+  val add_value_proj : t -> Ident.t -> shape -> t
```
I am wondering if it would be clearer to rename the `*_proj` family of function as `import_*`.

Drup commented:
First off: I'm very much in favour of having something like this. I longed for something similar in several projects (and even once looked at how jump-to-def was implemented in merlin to consider reusing it, and recoiled in fear). We might finally implement jump-to-def in `ocp-browser` by using those shapes ...

On the technical side: I think the shape language look reasonable. Is it already possible to use this to establish equivalence classes of symbols (by figuring out that `X.foo` is the same as `Y.bar` because `X` includes `Y`)?

> The compiler loads a lot of cmi files. Adding more to them might affect performance. I'd rather have the separate cmsi file, and consider it a replacement for cmti files in the usual dev build.

Could we test this experimentally? Given how small shapes seems to be, it seems like we could almost consider having them by default inside the `cm*` files ....

Octachron reviewed (commented):
Looking at the code itself, it seems pleasantly boring. Outside of outside of the questions on recursive modules and the name of the flag, I have some preliminary mid-level comments:

Octachron commented:
> One thing that took me a bit of time to understand is that shapes are abstractions of modules, not of module types.

That what was I meant by calling `cms` the `cmi file for module`, but my wording was utterly sibylline.

trefis commented:
If I'm not mistaken, it seems that people are generally OK with the idea of shapes and the way in which they have been implemented (though a formal approval is still missing), but are reluctant to introduce a new build artifact.

I will admit that I do not fully understand why (especially if you see it as a replacement for `.cmt` files), but will not press the issue right now.
However, I do not like the other suggestions which have been made:
- adding more things in `.cmi` seems wrong to me; and only solves half of the issue, we still need to store shapes for `.ml` files somewhere
- adding them in cmt files alongside a flag to prevent storing the typedtree in cmt files just seems like a bad idea. I am no build system expert so I might be making problems up where none exist, but as I said before: I think it's going to be problematic (despite @dbuenzli's remark that it's nothing new under the sun).  
The problem is not that the content of a file depends on how it was built, but that currently the build systems I know of can express things like "foo.odoc depends on foo.cmt", but not "foo.odoc depends on foo.cmt BUILT WITH[OUT] flag -whatever". (ping @jeremiedimino and @aalekseyev to confirm).

Anyway, one way to "solve" the issue for now, if people really do not want `.cms` files (though I'd like to hear why), would be to just add all the info we need to `.cmt[i]` files (making them even bigger! yeah!), and restart that discussion at a later date.
That would still leave us with shapes and an improved merlin, I'm happy to settle for that.

gasche commented:
I'm wary of the multiplication of artifacts as I think they have various costs/externalities (we need to train us to un-see them in build-artifact listings, we need to tell {users, build systems, editors} about them, we need tools to inspect them, etc.).

 That doesn't mean being opposed to the addition of new artifacts when they are well-motivated. In your case I understand the motivation for `.cms` files, but not really the motivation for `.cmsi` files (as far as I know, they don't actually contain shapes, despite the "s"). I would rather go for "one new artifact instead of two", but I think that a no-new-artifact approach would also make sense as a first iteration, to make the PR less invasive, and it should be easy to split the content as a new artifact later if we want to.

voodoos commented:
In 85eae5a I deleted the creation of the new `cms` file format and stored the shapes in the cmt files instead as this might be the most consensual solution right now.

(We can simply revert this commit if we decide otherwise)

xavierleroy commented:
> In 85eae5a I deleted the creation of the new cms file format and stored the shapes in the cmt files instead

Thank you.  Echoing @gasche's opinion with less fancy words: there's many things you can do to improve OCaml tooling, but adding a new kind of compiled file that needs to be installed is not one of them.

lpw25 commented:
As long as the shapes are also in `.cmt` files then you wouldn't have to install `.cms` files. They would simply be used during development within the repo you are working on.

As a slight variation on the original plan, how about a tool to extract the `.cms` from the `.cmt` along with merlin support for using `.cms` files when they are present? You don't save the time creating the `.cmt` file, but at least you save the disk space. (Also then I can just roll a version of the compiler that generates `.cms` and merlin will work with it out of the box).

trefis commented:
> Also then I can just roll a version of the compiler that generates `.cms` and merlin will work with it out of the box

Let us settle on that then, instead of adding an intermediate layer of tooling.
So, in conclusion, for now:
- the compiler stores shapes (and loc tables) in cmt files
- merlin looks for shapes in cms files (if they exist), or in cmt files
- janestreet has its own compiler fork which outputs cms files

And we'll revisit that decision after 5.0

Octachron commented on typing/shape.mli:
```diff
@@ -0,0 +1,153 @@
+    - an environment and a function to find shapes by path in that environment
+    - a function to load the shape of an external compilation unit
+    - some fuel, which is used to bound recursion when dealing with recursive
+      modules. (FTR: merlin currently uses a fuel of 10, which seems to be
```
I would suggest to replace `recursive modules` by `recursive shapes introduced by recursive modules` to make it more explicit that the fuel only matters in presence of recursive `include`s or functor applications in the implementation and not all recursive modules.

voodoos replied:
Good point, done in 348e7bb

Octachron commented on typing/typemod.ml:
```diff
@@ -1947,18 +1939,19 @@ let check_recmodule_inclusion env bindings =
         List.map
-          (fun (id, _name, _mty_decl, _modl, mty_actual, _attrs, _loc, _uid) ->
+          (fun (id, _name, _mty_decl, _modl,
+                mty_actual, _attrs, _loc, shape, _uid) ->
```
9-tuples are far beyond my personal limit to define a record, but this is orthogonal to this PR.

Octachron commented on typing/typemod.ml:
```diff
@@ -2154,53 +2170,63 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
+        wrap_constraint_with_shape env false md
+          (Mty_signature sg') shape Tmodtype_implicit
+      in
+      md, shape
```
This `let md, shape =... in md, shape` fragment is not necessary anymore with `wrap_constraint_with_shape`.

voodoos replied:
Done in f02ae2a

Octachron commented on .gitignore:
```diff
@@ -22,9 +22,9 @@
 *.dll
 *.la
-*.cm[ioxat]
+*.cm[ioxats]
```
This change can be removed for now.

voodoos replied:
Done in f02ae2a

Octachron commented on typing/typemod.ml:
```diff
@@ -1418,7 +1388,8 @@ and transl_signature env sg =
             List.iter (fun td ->
-              Signature_names.check_type names td.typ_loc td.typ_id
+              Signature_names.check_type names td.typ_loc td.typ_id;
+              Env.register_uid td.typ_type.type_uid td.typ_loc
```
There are also `...#row` private type in signature that should not be registered.

voodoos replied:
Done in 7b98372 and 53e313f

Octachron reviewed (approved):
I am approving the current `cmt` only version. By adding the shape data to the cmt (and the location table to cmti), this PR should be mostly invisible for unconcerned end-users while offering a simple data structure for any developer tool trying to inspect the content of implementation files.

The computation of shape itself is done by adding straightforward and well separated steps to the typing of module. Like stated by @gasche, the shape could be computed starting from the final typed tree, however computing the shape during typing makes most of the code obviously correct. From my point of view, this is a net positive since we want to decrease the total amount of work shared by the compiler and the related core tools (for instance merlin). Moreover, the well-separated nature of the shape computation code should make it easier to evolve (or remove if needed).

Octachron commented:
@voodoos : do you have any plans to clean/rewrite the commit history? Non-compiling commits are annoying when bisecting, and there is one more bootstrap than needed.

Octachron commented on typing/shape.ml:
```diff
@@ -238,41 +238,42 @@ module Make_reduce(Params : sig
+
+  let reduce env t =
+    fuel := Params.fuel;
+    reduce env t
```
I didn't notice this change when rereading the code, but the previous use of fuel wasn't incorrect: it is was limiting the depth of expansion of recursive module variables, whereas the current version is limiting the total number of expansions. Both strategies seem valid to me?

trefis replied:
I have the same understanding of the code as you, the reason for the change was that the previous version was looping on @gasche's example:
```ocaml
module rec A
  : sig val x : int -> int end
  = struct include B end
 and B
  : sig val x : int -> int end
  = struct include A end
```

voodoos commented:
> do you have any plans to clean/rewrite the commit history?

Yes I can do that. Is more than a big squash expected ?

Octachron commented:
If you are fine with just squashing the history, just tell me and I will squash the commits from the github interface. I was thinking that you might want to preserve some meaningful middle point commits.

voodoos committed:
Shapes: Add new tests and update others relying on idents

Co-authored-by: Thomas Refis <thomas.refis@gmail.com>

voodoos committed:
Shapes: Update env. to store shapes and uid/loc tables

Co-authored-by: Thomas Refis <thomas.refis@gmail.com>

voodoos committed:
Shapes: Add Changelog entry

voodoos commented:
I rebased with hopefully meaningful parts, but I still need to check that every commit builds.

@Octachron  When you say "Non-compiling commits are annoying" do you mean the `make world` target or a less extensive one ?

Octachron commented:
Ideally, `make world`, since it is hard to predict which kind of bug future bisect users will be investigating.
Your third commit doesn't compile (the changes in the toplevel reflect changes in `Typemod` that comes from the subsequent commit).

voodoos commented:
> the changes in the toplevel reflect changes in `Typemod` that comes from the subsequent commit

I switched them. 

However I am not able to build the commits separately (I get `inconsistent assumptions` errors even on the first one).

So feel free to squash if there is no easy way around. At least it will be clean on that branch as a reference.

Thanks again for you review !

voodoos committed:
Shapes: Store shapes in CMT* files

Co-authored-by: Thomas Refis <thomas.refis@gmail.com>

voodoos commented:
I rebased one more time. All commits now build independently (`make clean && make world`).

~~However to achieve that goal I had to make two successive bootstraps in commit 1 (98f040f) and 3 (e0ffed2). Is that ok or should I squash the first 3 commits together ?~~

Added `/!\ Needs boostrap` in commit messages and commited only one final bootstrap.

voodoos committed:
Shapes: Add Shape module with terms definitions and utility functions

/!\ Need bootstrap

+ Move Uid and Sig_component_kind to Shape

Co-authored-by: Thomas Refis <thomas.refis@gmail.com>

voodoos committed:
Shapes: Compute shapes during type checking

/!\ Needs bootstrap

Co-authored-by: Thomas Refis <thomas.refis@gmail.com>

voodoos committed:
Shapes: Add -dshape flag for debugging and testing purposes

Co-authored-by: Thomas Refis <thomas.refis@gmail.com>

voodoos committed:
Shapes: new bootstrap

trefis commented:
The commit history was clean, the CI was green, and there was an approval stamp as well as 4.14 milestone, so I just merged this on trunk (manually, to redo the bootstrap commit).

I'll cherry-pick (and squash) on 4.14 in a minute.

trefis commented:
Cherry-picked on 4.14 as 2a11a0a98c1e52c5c69a017f8da25c91eb017aaf .
